File: leetcode/tree/maximum_depth_binary_tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func solve(root *TreeNode, d int) int {
	if root == nil {
		return d
	}
	d++
	t1 := solve(root.Left, d)
	t2 := solve(root.Right, d)
	if t1 > t2 {
		return t1
	} else {
		return t2
	}
}

func maxDepth(root *TreeNode) int {
	return solve(root, 0)
}

func main() {
	root := &TreeNode{3, nil, nil}
	// root.Left = &TreeNode{9, nil, nil}
	root.Right = &TreeNode{20, nil, nil}
	root.Right.Left = &TreeNode{15, nil, nil}
	root.Right.Right = &TreeNode{7, nil, nil}
	depth := maxDepth(root)
	fmt.Printf("max depth: %d\n", depth)
}

File: leetcode/tree/maximum_depth_binary_tree.py
from common import TreeNode

def maxDepth(root: TreeNode) -> int:
    def solve(root, d):
        if not root:
            return d
        d += 1
        return max(solve(root.left, d), solve(root.right, d))
    return solve(root, 0)

if __name__ == "__main__":
    root = TreeNode(3)
    # root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print(maxDepth(root))